### PR TITLE
Sticky Defaults: Make Taho-as-default replace MetaMask in almost all cases

### LINKED ISSUE #2506: Tally Ho won't connect to purchase on Unstoppable Domains
### Discord Discussion Link

_No response_

### What browsers are you seeing the problem on?

Chrome

### What were you trying to do?

I was trying to connect my Tally Ho wallet to purchase a domain name on Unstoppable Domains. 

### What did not work?

Clicking on "MetaMask" nothing popped up. When I clicked on "Other Wallet" my MetaMask browser popped up even though I have Tally Ho set as default.
Screen Shot 2022-10-25 at 5 01 31 PM

### Version

v0.17.3

### Relevant log output

_No response_

### LINKED ISSUE #2706: Tally Ho unable to connect to BenQi
Not present when metamask is installed and enabled alongside Tally.

https://app.benqi.fi/markets

https://user-images.githubusercontent.com/94649004/205131923-045614b2-ad65-4b23-a387-0b6ef2acfa6b.mov

### LINKED ISSUE #2818: Connecting to zigzag exchange gives unexpected error
### Discord Discussion Link

n/a

### What browsers are you seeing the problem on?

Chrome

### What were you trying to do?

Connect to ZigZag exchange

### What did not work?

Extension pops up with an 'unexpected error' message

https://user-images.githubusercontent.com/71384922/209774768-0f226423-ce72-4cfb-b3a0-ea16db3cd12c.mov

### Version

v0.18.6

### Relevant log output

_No response_

### LINKED ISSUE #3231: Impersonate MM on starkgate.starknet.io
### Discord Discussion Link

https://discord.com/channels/808358975287722045/808358975287722048/1092385060889579621

### What would you like to be able to do?

https://starkgate.starknet.io/

Unable to connect to Taho on the Stark's official bridge

### LINKED ISSUE #3242: Impersonate MM
### Discord Discussion Link

https://discord.com/channels/808358975287722045/808358975287722048/1092491259458899969

### What would you like to be able to do?

https://router.via.exchange/

Unable to connect to Taho on the Via Protocol

### LINKED ISSUE #3368: Can't connect to bitcoinbridge.network/bridge
Adding this dApp to our MetaMask impersonation list is not enough to get it to work with Taho

### LINKED ISSUE #3412: Can't connect to Trader Joe if wallet is set as a default
Problem with connection on [Trader Joe](https://traderjoexyz.com/): 

If Taho is set as not default then it is listed as a Browser wallet option:
image

If Taho is set as default wallet then it should be available under Metamask option but it is no longer working. Browser wallet option is gone too.
image

### LINKED ISSUE #3433: I would like to connect/manage my Unstoppable Domain using the Taho Wallet
### Discord Discussion Link

https://discord.com/channels/808358975287722045/888500174685614090/1114974552632590416

### What browsers are you seeing the problem on?

Chrome

### What were you trying to do?

I own an Unstoppable Domain from https://unstoppabledomains.com/ Typically owners will 1. connect their wallet ( I used MEW Wallet) 2. claim their UD domain 3. then manage the domain including reverse resolution or 4. transfer the domain.
I have reached out to several members of the UD team, but I am not certain why we are not able to connect/manage with Taho.

I would love to help in anyway I can I think there are a lot of UD users that will benefit from our open source wallet, that are otherwise forced to use other wallets. 

### What did not work?

when I go to https://unstoppabledomains.com/ I am not able to connect. I have tried using Metamask, Wallet Connect, and "Other" options available on the UD page. I have not seen any errors, it is simply not recognizing our extension. 

### Version

v0.34.0

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by Shadowfiend**
## Background

Up until now, Taho's approach to being marked as default was meant to
be roughly:

- If Taho is marked as default, and the dApp supports showing a non-MetaMask
  provider at `window.ethereum`, make Taho that provider.
- If Taho is marked as default, and the dApp *doesn't* support showing a non
  MetaMask provider, have Taho impersonate MetaMask so that users can still
  connect using Taho.
- If Taho is marked as not-default, leave everything untouched, and only
  show Taho for dApps that specifically look for it (typically at `window.tally`,
  but sometimes in `window.ethereum.providers`).

For this to work, the window provider was tracking an allowlist of sites to
impersonate MetaMask on, and it was *also* tracking an allowlist of sites to
do additional tricks on (e.g. returning a blank `providers` array for sites
that detect MetaMask and don't look for any other wallet). There were other
tricks also being done to try to discern between scenarios, etc.

Other wallets just throw their hands up and attempt to impersonate MetaMask
most/all the time, leading to escalating wars with dApp frameworks to detect
these behaviors.

## Revised Approach

This PR throws all of that away. It was confusing for users (do you click on
MetaMask? Browser/Injected? What happens if Taho isn't default? What about
the times when MetaMask impersonation isn't hardcoded for a site?), and it
was difficult to maintain (allowlists of specific sites are not realistically
a scalable long-term approach).

Instead, having Taho set as default (a reminder that as of #3492, this is
represented in the UI as a toggle between “connect with MetaMask” and “connect
with Taho”) should mean two things:

- Trying to connect to MetaMask will *always* use Taho.
- For dApps that also specifically detect Taho, they may *also* allow connecting
  with Taho. If Taho is set as default, attempting to connect with MetaMask will
  still connect with Taho.

On the flip side, if Taho is not set as default, it should mean two things:

- Trying to connect to MetaMask will *always* use whatever was presenting
  as MetaMask before Taho got involved.
- For dApps that also specifically detect Taho, they will allow connecting
  with Taho still.

### Technical Details

How we do this is straightforward: any provider on the `providers` array
that presents as MetaMask (our heuristic: `isMetaMask` returns `true` AND
there are no other `is*` methods on the provider) is wrapped in a one-time
proxy. That proxy passes everything through to the wrapped provider, *unless*
Taho has been set to present as MetaMask by the user (an explicit choice).
In this case, a subset of functions are always redirected to Taho, including
account and RPC requests. Additionally, any calls to `isMetaMask` on
`window.ethereum` return `true` if Taho is set to default, even though the
provider presented on `window.ethereum` is the Taho provider.

We do two more small things:
- When Taho is set as default, it is set as the first provider in the providers
  array. When it is not, it is removed from the front of the providers list. This
  allows MetaMask or other wallet usage for dApps that use a framework like wagmi,
  which especially in older versions only used the first wallet in the providers
  array for connections. Examples of dApps that show this behavior include Lens
  and Vela (e.g. see [this bug report in Discord](https://discord.com/channels/808358975287722045/888500174685614090/1127318749746311258)).
- When Taho is set as default by the user and MetaMask is not installed, a mock
  MetaMask is presented to the dApp that reroutes calls to Taho. If Taho is
  switched back to not default, the mock is removed. This allows dApps like
  bitcoinbridge.network that only allow connecting via MetaMask to work with
  Taho even if MetaMask is not installed.

The net result is that the “act as MetaMask”/“Taho as default” toggle should
behave consistently across all dApps. Famous last words :)

Finally, we fix a small issue with eth_requestAccounts handling on certain dApps.
Not related to MetaMask impersonation really, but knockin' bugs out.

## Testing

- This is going to require doing some relatively extensive testing in our
  release QA dApps to make sure that MetaMask is replaced when Taho is set
  as default and not when
  it is not.
- All these should also be replicated with Taho as default and not and MetaMask
  disabled (to ensure the dApp registers MetaMask as uninstalled when Taho is
  not default, and allows interaction as MetaMask when Taho is default).
- We should also check Lens (https://claim.lens.xyz/) to make sure the Browser
  connection uses Taho when default, and MetaMask when not.
- Finally, the dApps in the issues below should be tested to make sure they're
  all working!

Fixes #3433, fixes #3412, fixes #3368, fixes #3242, fixes #3231, fixes #2506, fixes #2706, fixes #2818.

Latest build: [extension-builds-3546](https://github.com/tahowallet/extension/suites/14241880950/artifacts/799754754) (as of Wed, 12 Jul 2023 12:06:52 GMT).

### CODE REVIEW

**Review by jagodarybacka — COMMENTED** [review]
Looks good, will take a second look at it tomorrow

**Review by kkosiorowska — COMMENTED** [review]
Code looks good, I'm getting down to testing.

**Comment by kkosiorowska** [comment]
Tests notes:

Following issues:
- [X] https://github.com/tahowallet/extension/issues/3412
- [X] https://github.com/tahowallet/extension/issues/3368
- [X] https://github.com/tahowallet/extension/issues/3242
- [X] https://github.com/tahowallet/extension/issues/3231
- [X] https://github.com/tahowallet/extension/issues/2506 and https://github.com/tahowallet/extension/issues/3433
- [ ] https://github.com/tahowallet/extension/issues/2706
I managed to connect, but the page keeps reloading.
https://github.com/tahowallet/extension/assets/23117945/c1cabffb-bc68-45b1-9e6a-8cc332c7f907
- [ ] https://github.com/tahowallet/extension/issues/2818
The error has disappeared but I'm not able to connect further. 
https://github.com/tahowallet/extension/assets/23117945/1300de3b-c49a-4079-83e5-e77f22d06fa2


dApps:
- [X] https://claim.lens.xyz/
- [x] https://matcha.xyz/
- [x] https://wallet.polygon.technology/
- [x] https://galxe.com/
- [x] https://web3modal.com/
- [x] https://curve.fi/#/ethereum/swap

**Review by kkosiorowska — APPROVED** [review]
Everything seems to work well with the exception of two issues (https://github.com/tahowallet/extension/issues/2706 and https://github.com/tahowallet/extension/issues/2818). Let's not block this and include it in the upcoming release because it makes for a big bang improvement for our users. 🚀

File: window-provider/index.ts [hunk 1]
```diff
@@ -141,7 +143,7 @@ export default class TallyWindowProvider extends EventEmitter {
       }
 
       if (isTallyConfigPayload(result)) {
-        const wasTallySetAsDefault = this.tallySetAsDefault
+        const wasTallySetAsDefault = this.tahoSetAsDefault
```
[the remark this answers is not in the document]
Comment (kkosiorowska) [reply window-provider/index.ts:146]: Yes, I agree. Let's do it next steps.

**Pull request merged**